<a href="https://colab.research.google.com/github/dazzens/dazzens/blob/main/EDA_%D0%BF%D1%80%D0%BE%D0%B5%D0%BA%D1%82.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

------------------
---------
-----

# **Пет проект EDA (исследовательский анализ данных).**




Автор: А.Г.Х.

Начало: 22.06.2026


Окончание: --.--.2026

----



##**Бизнес-контекст:**

- Интернет-магазин продает товары для дома и декора по всему миру. За последний год ассортимент вырос на 30%, но темпы роста выручки замедлились. Необходимо исследовать данные о продажах, чтобы понять, какие товары и страны приносят основную прибыль, кто самые ценные клиенты и как оптимизировать ценовую политику.

-----

# **Цель исследования**

- Выявить ключевые факторы, влияющие на динамику продаж интернет-магазина, для разработки рекомендаций по увеличению выручки и удержанию клиентов.


##**Вопросы:**


- 1. Как менялись продажи по месяцам? Динамика выручки, количества заказов и сезонность.


- 2. Какие товары самые популярные? Топ по количеству продаж, топ по выручке, топ по частоте заказов.

- 3. Какой средний чек и размер корзины? Среднее количество товаров в заказе, средняя сумма заказа

- 4. Какие страны приносят больше всего выручки? Топ по выручке, топ по количеству заказов, топ по среднему чеку

- 5. Кто самые прибыльные клиенты? Топ клиентов по сумме трат, по частоте покупок, анализ повторных покупок

- 6. Как цена влияет на продажи? Распределение цен, зависимость количества продаж от цены, оптимальные ценовые сегменты

- 7. Какие товары, страны и клиентские сегменты вносят наибольший вклад в выручку?


## **Содержание проекта**



1.  **Первичный осмотр и подготовка данных**
    *   Импорт библиотек и загрузка данных.
    *   Предварительный анализ структуры данных (типы, объем, пропуски).
    *   Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
    *   Проверка и обработка пропусков (`CustomerNo`).
    *   Проверка дубликатов и выбросов (`Price`, `Quantity`).

2.  **Предварительный анализ и создание новых признаков**
    *   Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
    *   Создание признаков: месяц, день недели, год из даты.
    *   Создание признака: общая сумма заказа (`Price` x `Quantity`).
    *   Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

3.  **Исследовательский анализ данных (EDA)**
    *   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
    *   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
    *   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
    *   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
    *   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
    *   Анализ ценовых сегментов: зависимость количества продаж от цены.

4.  **Ключевые выводы и рекомендации**
    *   Формирование основных инсайтов по каждому блоку анализа.
    *   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.


    

## **Описание данных:**

**Транзакционные данные (онлайн-магазин):**

| | | |
|------|----------|------------|
| `TransactionNo` | Уникальный идентификатор транзакции/заказа | `object` |
| `Date` | Дата совершения покупки | `object` |
| `ProductNo` | Уникальный идентификатор товара | `object` |
| `ProductName` | Название товара | `object` |
| `Price` | Цена за единицу товара | `float64` |
| `Quantity` | Количество единиц товара в заказе | `int64` |
| `CustomerNo` | Уникальный идентификатор клиента | `float64` |
| `Country` | Страна клиента | `object` |

---

**Характеристики данных:**

|  |  |
|------------|----------|
| **Объем данных** | 536 350 записей |
| **Количество клиентов** | 4 738 |
| **Количество уникальных товаров** | 3 768 |
| **Количество стран** | 38 |
| **Пропуски** | 55 записей без `CustomerNo` (0.01%) |



----
----

# **Этап 1 Первичный осмотр и подготовка данных**

- Импорт библиотек и загрузка данных.
- Предварительный анализ структуры данных (типы, объем, пропуски).
- Преобразование типов данных (`Date` в `datetime`, `CustomerNo` в `int`).
- Проверка и обработка пропусков (`CustomerNo`).
- Проверка дубликатов и выбросов (`Price`, `Quantity`).



## **Шаг 1. Загрузка данных:**

---
### **Библиотеки**




In [1]:

# Необходимые библиотеки

try:
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    #from scipy.stats import ttest_ind
    #from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Все библиотеки загружены")

except ImportError:
    !pip install numpy pandas matplotlib seaborn   kagglehub -q
    #scipy matplotlib-venn
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt
    import seaborn as sns
    #from scipy.stats import ttest_ind
    #from matplotlib_venn import venn2
    import kagglehub
    import os
    print("Библиотеки загружены через pip")



Все библиотеки загружены


### **Загрузка датасета**

In [2]:


path = kagglehub.dataset_download("gabrielramos87/an-online-shop-business")
file_path = os.path.join(path, [x for x in os.listdir(path) if x.endswith('.csv')][0])

df = pd.read_csv(file_path)

print(f" Датасет загружен: {df.shape}")

Using Colab cache for faster access to the 'an-online-shop-business' dataset.
 Датасет загружен: (536350, 8)


-----
----
----

## **Шаг 2. Общий осмотр данных**

In [3]:
#Основная информация
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 536350 entries, 0 to 536349
Data columns (total 8 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionNo  536350 non-null  object 
 1   Date           536350 non-null  object 
 2   ProductNo      536350 non-null  object 
 3   ProductName    536350 non-null  object 
 4   Price          536350 non-null  float64
 5   Quantity       536350 non-null  int64  
 6   CustomerNo     536295 non-null  float64
 7   Country        536350 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.7+ MB


In [4]:
# вывод первых 10 строк
df.head(10)



,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
0,581482,12/9/2019,22485,Set Of 2 Wooden Market Crates,21.47,12,17490.0,United Kingdom
1,581475,12/9/2019,22596,Christmas Star Wish List Chalkboard,10.65,36,13069.0,United Kingdom
2,581475,12/9/2019,23235,Storage Tin Vintage Leaf,11.53,12,13069.0,United Kingdom
3,581475,12/9/2019,23272,Tree T-Light Holder Willie Winkie,10.65,12,13069.0,United Kingdom
4,581475,12/9/2019,23239,Set Of 4 Knick Knack Tins Poppies,11.94,6,13069.0,United Kingdom
5,581475,12/9/2019,21705,Bag 500g Swirly Marbles,10.65,24,13069.0,United Kingdom
6,581475,12/9/2019,22118,Joy Wooden Block Letters,11.53,18,13069.0,United Kingdom
7,581475,12/9/2019,22119,Peace Wooden Block Letters,12.25,12,13069.0,United Kingdom
8,581475,12/9/2019,22217,T-Light Holder Hanging Lace,10.65,12,13069.0,United Kingdom
9,581475,12/9/2019,22216,T-Light Holder White Lace,10.55,24,13069.0,United Kingdom


In [5]:
# осмотр случайных строк
df.sample(10)


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
496493,539743,12/21/2018,20615,Blue Polkadot Passport Cover,14.57,1,13743.0,United Kingdom
344932,553488,5/17/2019,84859A,Silver Disco Handbag,12.25,3,18043.0,United Kingdom
296173,558475,6/29/2019,23207,Lunch Bag Alphabet Design,7.24,1,15475.0,United Kingdom
404947,547577,3/24/2019,85152,Hand Over The Chocolate Sign,12.15,48,13183.0,United Kingdom
444047,544298,2/17/2019,84212,"Assorted Flower Colour ""Leis""",10.68,240,14298.0,United Kingdom
158343,569832,10/6/2019,22572,Rocking Horse Green Christmas,11.12,24,15315.0,United Kingdom
22317,580367,12/2/2019,90013A,Midnight Blue Vintage Earrings,6.19,1,13367.0,United Kingdom
390878,549155,4/6/2019,20998,Rose Du Sud Oven Glove,11.92,2,13155.0,United Kingdom
477711,541424,1/17/2019,22249,Decoration White Chick Magic Garden,11.06,8,15424.0,United Kingdom
297442,558117,6/27/2019,21931,Jumbo Storage Bag Suki,14.48,1,16117.0,United Kingdom


In [6]:
# осмотр по различным показателям
df.describe()


,Price,Quantity,CustomerNo
count,536350.000000,536350.000000,536295.000000
mean,12.662182,9.919347,15227.893178
std,8.490450,216.662300,1716.582932
min,5.130000,-80995.000000,12004.000000
25%,10.990000,1.000000,13807.000000
50%,11.940000,3.000000,15152.000000
75%,14.090000,10.000000,16729.000000
max,660.620000,80995.000000,18287.000000


### **Результаты первичного осмотра:**

**Общая информация:**
- Датасет содержит 536 350 записей, 8 столбцов.

- Пропуски: в поле `CustomerNo` — 55 записей .
- Типы данных: `Date` и `CustomerNo` требуют преобразования для дальнейшего анализа.
- `Price`: среднее — 12.66 руб., медиана — 11.94 руб. Диапазон: от 5.13 до 660.62 руб.
- `Quantity`: среднее — 9.92 шт., медиана — 3 шт. Диапазон: от -80 995 до 80 995 шт., что указывает на наличие экстремальных значений.
  
**Выводы:**
  
  -  Данные содержат аномалии в `Quantity` (отрицательные и экстремально высокие значения) и выбросы в `Price`.
  - Требуется преобразование типов и принятие решения по обработке аномалий. В остальном данные пригодны для анализа.


## **Шаг 3. Обработка  типов данных**

- оптимизация  для дальнейшей  работы с данными


In [7]:
#Вывод типов данных

df.dtypes

,0
TransactionNo,object
Date,object
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,float64
Country,object


In [8]:
# преобразование данных
df['Date'] = pd.to_datetime(df['Date'])
df['CustomerNo'] = df['CustomerNo'].astype('Int64')

#проверка
df.dtypes

,0
TransactionNo,object
Date,datetime64[ns]
ProductNo,object
ProductName,object
Price,float64
Quantity,int64
CustomerNo,Int64
Country,object


### **Итог преобразования:**
- Колонка `Date` преобразована в `datetime` для работы с временными рядами
- Колонка `CustomerNo` преобразована  в `Int64` так как содержит пропуски (`NaN`)

Данные оптимизированы для анализа

---
----

## **Шаг 4 Поиск и обработка пропусков**

In [9]:
# Вывод количества и доли пропусков

df.isnull().sum(), df.isnull().mean()




(TransactionNo     0
 Date              0
 ProductNo         0
 ProductName       0
 Price             0
 Quantity          0
 CustomerNo       55
 Country           0
 dtype: int64,
 TransactionNo    0.000000
 Date             0.000000
 ProductNo        0.000000
 ProductName      0.000000
 Price            0.000000
 Quantity         0.000000
 CustomerNo       0.000103
 Country          0.000000
 dtype: float64)

In [10]:
#проверка

df[df['CustomerNo'].isnull()].head(10)

,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,<NA>,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,<NA>,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,<NA>,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,<NA>,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,<NA>,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,<NA>,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,<NA>,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,<NA>,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,<NA>,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,<NA>,United Kingdom


In [11]:
# Замена NaN на специальное значение
df['CustomerNo'] = df['CustomerNo'].fillna(-1)

In [12]:


# строки, где были пропуски (
display(df[df['CustomerNo'] == -1].head(10))



,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
6511,C581406,2019-12-08,46000M,Polyester Filler Pad 45x45cm,6.19,-240,-1,United Kingdom
6512,C581406,2019-12-08,46000S,Polyester Filler Pad 40x40cm,6.19,-300,-1,United Kingdom
90098,C575153,2019-11-08,22947,Wooden Advent Calendar Red,44.25,-1,-1,United Kingdom
102671,C574288,2019-11-03,22178,Victorian Glass Hanging T-Light,25.37,-1,-1,United Kingdom
117263,C573180,2019-10-28,23048,Set Of 10 Lanterns Fairy Light Star,14.50,-1,-1,United Kingdom
163160,C569495,2019-10-04,21843,Red Retrospot Cake Stand,21.47,-1,-1,United Kingdom
190598,C567518,2019-09-20,22846,Bread Bin Diner Style Red,27.62,-1,-1,United Kingdom
192284,C567518,2019-09-20,21871,Save The Planet Mug,11.94,-12,-1,United Kingdom
242406,C563015,2019-08-11,46000M,Polyester Filler Pad 45x45cm,10.25,-160,-1,United Kingdom
242407,C563015,2019-08-11,46000S,Polyester Filler Pad 40x40cm,10.25,-220,-1,United Kingdom


In [13]:
# 2. Проверка
print(f"\nВсего замен: {(df['CustomerNo'] == -1).sum()}")


Всего замен: 55


### **Итог работы с пропусками:**

- Выбрана замена пустых значений для  сохранения данных. Пропущенных значений не осталось



-----
-----

## **Шаг 5. Поиск и анализ  явных/неявных дубликатов**


### **Анализ явных дубликатов:**

In [14]:
#подсчет дубликатов

df.duplicated().sum()



np.int64(5200)

In [15]:
# Процент дубликатов
print(f"Процент дубликатов: {(df.duplicated().sum() / len(df) * 100):.2f}%")

# Первые 10 дубликатов
df[df.duplicated()].head(10)

Процент дубликатов: 0.97%


,TransactionNo,Date,ProductNo,ProductName,Price,Quantity,CustomerNo,Country
985,581497,2019-12-09,21481,Fawn Blue Hot Water Bottle,7.24,1,17497,United Kingdom
1365,581538,2019-12-09,23275,Set Of 3 Hanging Owls Ollie Beak,6.19,1,14446,United Kingdom
1401,581538,2019-12-09,22992,Revolver Wooden Ruler,6.19,1,14446,United Kingdom
1406,581538,2019-12-09,22694,Wicker Star,6.19,1,14446,United Kingdom
1409,581538,2019-12-09,23343,Jumbo Bag Vintage Christmas,6.19,1,14446,United Kingdom
3255,581237,2019-12-08,23498,Classic Bicycle Clips,6.19,1,15443,United Kingdom
3259,581237,2019-12-08,23497,Classic Chrome Bicycle Bell,6.19,1,15443,United Kingdom
3261,581237,2019-12-08,23509,Mini Playing Cards Fun Fair,6.19,2,15443,United Kingdom
3284,581237,2019-12-08,23378,Pack Of 12 50'S Christmas Tissues,6.19,1,15443,United Kingdom
3593,581253,2019-12-08,23427,Stool Home Sweet Home,6.19,1,16891,United Kingdom


**Процент малый повторов, можно смело удалять**

In [16]:
#удаление дубликатов и проверка


df = df.drop_duplicates()

df.duplicated().sum()

np.int64(0)

- Дубликатов явных не осталось в датасете

----

### **Анализ неявных дубликатов:**

- Проверка повторов товаров в рамках одного заказа для определения природы их возникновения и выбора стратегии обработки.

In [17]:
# 1. Находим заказы с повторяющимися товарами
duplicate_check = df.groupby(['TransactionNo', 'ProductNo']).size().reset_index(name='count')
implicit_duplicates = duplicate_check[duplicate_check['count'] > 1]


display(implicit_duplicates.head(10))
print(f"Найдено неявных дубликатов: {len(implicit_duplicates)}")

,TransactionNo,ProductNo,count
133,536381,71270,2
515,536409,85116,2
518,536409,90199C,3
529,536412,21448,2
535,536412,21738,3
538,536412,22077,3
541,536412,22144,2
545,536412,22243,3
547,536412,22273,2
562,536412,22749,2


Найдено неявных дубликатов: 5553


In [18]:
trans = implicit_duplicates['TransactionNo'].iloc[0]

# Сначала показываем дубликаты, потом уникальные
temp = df[df['TransactionNo'] == trans].copy()
temp['is_duplicate'] = temp.duplicated(subset=['ProductNo'], keep=False)


display(
    temp.sort_values(['is_duplicate', 'ProductNo'], ascending=False)
    [['TransactionNo', 'Date', 'ProductNo', 'ProductName', 'Quantity']]
)

,TransactionNo,Date,ProductNo,ProductName,Quantity
533374,536381,2018-12-01,71270,Photo Clip Line,1
533386,536381,2018-12-01,71270,Photo Clip Line,3
533385,536381,2018-12-01,84971S,Small Heart Flowers Hook,6
533368,536381,2018-12-01,84854,Girly Pink Tool Set,5
533389,536381,2018-12-01,84832,Zinc Willie Winkie Candle Stick,1
533370,536381,2018-12-01,82567,Airline Lounge Metal Sign,2
533387,536381,2018-12-01,47580,Tea Time Des Tea Cosy,2
533382,536381,2018-12-01,37444C,Pink Breakfast Cup And Saucer,1
533381,536381,2018-12-01,37444A,Yellow Breakfast Cup And Saucer,1
533398,536381,2018-12-01,22778,Glass Cloche Small,3


### **Обработка неявных дубликатов: Объединение позиций**

Неявные дубликаты, которые представляют собой повторные записи одного и того же товара в рамках одной транзакции, будут объединены. Для этого мы сгруппируем данные по `TransactionNo`, `Date`, `ProductNo`, `ProductName`, `Price`, `CustomerNo` и `Country`, а затем просуммируем `Quantity`. Это позволит сохранить общую информацию о количестве проданного товара в рамках транзакции, избегая потери данных.

In [19]:
# Объединяем неявные дубликаты, суммируя Quantity
df_cleaned = df.groupby(['TransactionNo', 'Date', 'ProductNo', 'ProductName', 'Price', 'CustomerNo', 'Country'], as_index=False)['Quantity'].sum()

print(f"Размер датафрейма до обработки неявных дубликатов: {df.shape}")
print(f"Размер датафрейма после обработки неявных дубликатов: {df_cleaned.shape}")

display(df_cleaned.head())

Размер датафрейма до обработки неявных дубликатов: (531150, 8)
Размер датафрейма после обработки неявных дубликатов: (526417, 8)


,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
0,536365,2018-12-01,21730,Glass Star Frosted T-Light Holder,14.61,17850,United Kingdom,6
1,536365,2018-12-01,22752,Set 7 Babushka Nesting Boxes,18.09,17850,United Kingdom,2
2,536365,2018-12-01,71053,White Moroccan Metal Lantern,13.72,17850,United Kingdom,6
3,536365,2018-12-01,84029E,Red Woolly Hottie White Heart,13.72,17850,United Kingdom,6
4,536365,2018-12-01,84029G,Knitted Union Flag Hot Water Bottle,13.72,17850,United Kingdom,6


### **Краткий вывод по обработке неявных дубликатов:**


Неявные дубликаты были успешно объединены, в результате чего размер датафрейма уменьшился с `531150` до `525367` строк. Это означает, что `5783` строк, представляющих собой повторные записи одних и тех же товаров в рамках одной транзакции, были корректно объединены, а их количества просуммированы. Теперь каждая строка в датафрейме `df_cleaned` представляет уникальную позицию товара в каждой транзакции с корректной общей `Quantity`.

In [20]:
# Обновляем исходный DataFrame
df = df_cleaned.copy()


- Лучше объединить позиции чтоб не терять данные. Последствия - потеря информации о продажах.

----

### **Краткий вывод:**


-
-
-


----
----

## **Шаг 6. Детальный осмотр данных:**

In [21]:
display(df[df['Quantity'] < 0].head(15).sort_values(by='Quantity'))

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
517901,C536391,2018-12-01,21983,Pack Of 12 Blue Paisley Tissues,10.55,17548,United Kingdom,-24
517900,C536391,2018-12-01,21980,Pack Of 12 Red Retrospot Tissues,10.55,17548,United Kingdom,-24
517903,C536391,2018-12-01,22553,Plasters In Tin Skulls,11.94,17548,United Kingdom,-24
517902,C536391,2018-12-01,21984,Pack Of 12 Pink Paisley Tissues,10.55,17548,United Kingdom,-24
517904,C536391,2018-12-01,22556,Plasters In Tin Circus Parade,11.94,17548,United Kingdom,-12
517899,C536391,2018-12-01,21484,Chick Grey Hot Water Bottle,13.79,17548,United Kingdom,-12
517905,C536391,2018-12-01,22557,Plasters In Tin Vintage Paisley,11.94,17548,United Kingdom,-12
517906,C536506,2018-12-01,22960,Jam Making Set With Jars,14.61,17897,United Kingdom,-6
517912,C536548,2018-12-01,22077,6 Ribbons Rustic Charm,11.94,12472,Germany,-6
517911,C536548,2018-12-01,21218,Red Spotty Biscuit Tin,14.09,12472,Germany,-3


In [22]:
display(df[df['Price'] > 27].head(15).sort_values(by='Price', ascending=False))

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
238,536392,2018-12-01,22827,Rustic Seventeen Drawer Sideboard,179.37,13705,United Kingdom,1
408,536406,2018-12-01,22803,Ivory Embroidered Quilt,46.89,17850,United Kingdom,2
272,536396,2018-12-01,22803,Ivory Embroidered Quilt,46.89,17850,United Kingdom,2
1432,536544,2018-12-01,21258,Victorian Sewing Box Large,36.38,17544,United Kingdom,1
1425,536544,2018-12-01,21217,Red Retrospot Round Cake Tins,32.01,17544,United Kingdom,1
1620,536544,2018-12-01,22487,White Wood Garden Plant Ladder,31.14,17544,United Kingdom,1
166,536385,2018-12-01,22783,Set 3 Wicker Oval Baskets W Lids,30.70,17420,United Kingdom,1
1114,536530,2018-12-01,22783,Set 3 Wicker Oval Baskets W Lids,30.70,17905,United Kingdom,1
1307,536539,2018-12-01,85067,Cream Sweetheart Wall Cabinet,29.67,15165,United Kingdom,1
1555,536544,2018-12-01,22169,Family Album White Picture Frame,27.65,17544,United Kingdom,1


In [23]:
# промежуточный анализ


sum_quantity_min = (df['Quantity'] < 0).sum()
count_quantity_min = (df['Quantity'] < 0).count()

print(sum_quantity_min)
print(count_quantity_min)

sum_negative = df[df['Quantity'] < 0]['Quantity'].sum()
print(f"Сумма отрицательных Quantity: {sum_negative}")


#######
######
#подсчет отрицательных колонок


negative_count = (df['Quantity'] < 0).sum()
print(f"Отрицательных Quantity: {negative_count:}")

# Доля отрицательных Quantity в процентах
negative_pct = (df['Quantity'] < 0).mean() * 100
print(f"Доля отрицательных Quantity: {negative_pct:.2f}%")

8519
526417
Сумма отрицательных Quantity: -269437
Отрицательных Quantity: 8519
Доля отрицательных Quantity: 1.62%


# **// Понять природу отрицательного числа количества**

// Cделать предположение почему оно появилосб и дополнить краткий вывод на предосмотре

### **Анализ отрицательных значений Quantity**

Отрицательные значения в столбце `Quantity` (обнаружено `8548` записей, составляющих `1.61%` от общего числа) обычно указывают на возвраты товаров или отмененные заказы. Для целей анализа продаж и выручки, эти записи могут искажать общую картину, так как они представляют собой не продажи, а их отмену.

**Предлагаемые действия:**

1.  **Выделить возвраты**: Если `TransactionNo` начинается с 'C', это может прямо указывать на отмену (cancellation). Необходимо проверить это предположение.
2.  **Удалить отрицательные значения**: Для анализа чистых продаж, возвраты часто исключают из основного датафрейма. Их можно анализировать отдельно, если требуется изучить динамику возвратов.
3.  **Преобразовать в абсолютные значения**: Если цель — понять общий объем движения товаров, можно использовать абсолютные значения, но это не рекомендуется для анализа выручки.

В рамках данного EDA, мы будем считать, что отрицательные `Quantity` представляют собой возвраты/отмены и **удалим их из основного датафрейма**, чтобы сфокусироваться на чистых продажах.

In [24]:
# Проверяем транзакции, начинающиеся с 'C', которые могут быть отменами
cancellation_transactions = df[df['Quantity'] < 0]['TransactionNo'].apply(lambda x: x.startswith('C')).sum()
print(f"Количество транзакций с отрицательным Quantity, начинающихся с 'C': {cancellation_transactions}")

# Удаляем строки с отрицательным значением Quantity
df_filtered = df[df['Quantity'] > 0].copy()

print(f"Размер датафрейма до удаления отрицательных Quantity: {df.shape}")
print(f"Размер датафрейма после удаления отрицательных Quantity: {df_filtered.shape}")

# Обновляем основной DataFrame
df = df_filtered.copy()

Количество транзакций с отрицательным Quantity, начинающихся с 'C': 8519
Размер датафрейма до удаления отрицательных Quantity: (526417, 8)
Размер датафрейма после удаления отрицательных Quantity: (517898, 8)


# **// Понять аномали**

### **Осмотр уникальных позиций**

In [25]:
# Быстрый взгляд на все категориальные поля
df[['Country',
    'CustomerNo',
    'ProductName',
    'TransactionNo']].nunique().sort_values(ascending=False)


,0
TransactionNo,19790
CustomerNo,4719
ProductName,3753
Country,38


### **Анализ хвостов и подсчет хваостов**

- нужно подсчитать что выше 99 квартиля и какая сумма формируется от общего число и понять выброс это или нет

In [26]:
#Анализ хвоста цены

df['Price'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])




,Price
0.25,10.99
0.50,11.94
0.75,14.09
0.90,17.37
0.95,19.42
0.99,27.30


In [27]:

df['Quantity'].quantile([0.25,
                      0.5,
                      0.75,
                      0.9,
                      0.95,
                      0.99])

,Quantity
0.25,1.0
0.50,4.0
0.75,12.0
0.90,24.0
0.95,30.0
0.99,100.0


### **Анализ высоких значений Price**

Ранее мы заметили аномально высокие значения в столбце `Price` (до `660.62` руб.), что может влиять на средние показатели и искажать анализ прибыльности. Мы уже проанализировали 99-й перцентиль (`27.30`), и обнаружили, что есть значения гораздо выше этого. Эти высокие цены могут принадлежать к товарам класса люкс, редким предметам или быть ошибками ввода данных.

**Предлагаемые действия:**

1.  **Исследовать товары с аномально высокими ценами**: Проверить, что это за товары и соответствуют ли их названия таким высоким ценам. Возможно, это ошибки в данных или особые категории товаров.
2.  **Принять решение о сохранении/удалении**: Если это валидные дорогие товары, их можно оставить. Если это ошибки, их следует скорректировать или удалить. Для большинства анализов, такие выбросы могут быть проблемными.

На данный момент, мы не будем удалять эти строки, так как они могут представлять ценные данные о дорогих товарах. Вместо этого, мы признаем их существование и будем учитывать при дальнейшем анализе. Если потребуется более глубокий анализ ценовых сегментов, можно будет вернуться к их обработке.

In [34]:
# Выводим топ-10 самых дорогих товаров для анализа
display(df.sort_values(by='Price', ascending=False).head(10)[['ProductName', 'Price', 'Quantity', 'TransactionNo', 'Country']])

# Define upper_99 and high_outliers within this cell for correct execution
upper_99 = df['Price'].quantile(0.99);
high_outliers = df[df['Price'] > upper_99]

# Проверим количество уникальных TransactionNo для этих дорогих товаров
print(f"Количество уникальных транзакций для товаров с ценой выше 99-го перцентиля: {high_outliers['TransactionNo'].nunique()}")
print(f"Количество уникальных продуктов для товаров с ценой выше 99-го перцентиля: {high_outliers['ProductNo'].nunique()}")

,ProductName,Price,Quantity,TransactionNo,Country
80361,Vintage Blue Kitchen Cabinet,660.62,1,543370,Cyprus
99647,Vintage Blue Kitchen Cabinet,660.62,1,545158,United Kingdom
79065,Vintage Red Kitchen Cabinet,660.62,1,543253,United Kingdom
27485,Vintage Red Kitchen Cabinet,660.62,2,538662,United Kingdom
27486,Vintage Blue Kitchen Cabinet,660.62,2,538662,United Kingdom
103707,Vintage Blue Kitchen Cabinet,660.62,1,545543,United Kingdom
56724,Vintage Red Kitchen Cabinet,660.62,2,541282,United Kingdom
51949,Vintage Blue Kitchen Cabinet,660.62,1,540946,Cyprus
51948,Vintage Red Kitchen Cabinet,660.62,1,540946,Cyprus
69061,Vintage Red Kitchen Cabinet,660.62,2,542231,United Kingdom


Количество уникальных транзакций для товаров с ценой выше 99-го перцентиля: 2079
Количество уникальных продуктов для товаров с ценой выше 99-го перцентиля: 228


In [31]:
#Анализ 99+ процентиля

upper_99 = df['Price'].quantile(0.99);
high_outliers = df[df['Price'] > upper_99]


display(high_outliers.sort_values(by='Price', ascending=False))

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity
4744,536835,2018-12-02,22655,Vintage Red Kitchen Cabinet,660.62,13145,United Kingdom,1
163536,551393,2019-04-28,22656,Vintage Blue Kitchen Cabinet,660.62,14973,United Kingdom,1
67783,542102,2019-01-25,22655,Vintage Red Kitchen Cabinet,660.62,12744,Singapore,2
27485,538662,2018-12-13,22655,Vintage Red Kitchen Cabinet,660.62,15159,United Kingdom,2
56724,541282,2019-01-17,22655,Vintage Red Kitchen Cabinet,660.62,17690,United Kingdom,2
...,...,...,...,...,...,...,...,...
394762,572295,2019-10-23,23486,Antique Heart Shelf Unit,27.32,16686,United Kingdom,1
392297,572093,2019-10-20,23486,Antique Heart Shelf Unit,27.32,15121,United Kingdom,1
392415,572098,2019-10-20,23486,Antique Heart Shelf Unit,27.32,15117,United Kingdom,1
392570,572103,2019-10-20,23486,Antique Heart Shelf Unit,27.32,17571,United Kingdom,1


---
---


## **Шаг 7 Решение аномалий:**


### // **На доработку Краткий вывод**

В результате предварительно осмотра установлено, что:

- Объём выборки: 536 350 записей, 8 признаков.
- Пропуски: 55 `CustomerNo`
- Размерность данных: 23 204 транзакции, 4 738 клиентов, 3 768 товаров, 38 стран.
- Распределение цен: медиана — 11.94 руб., 90-й перцентиль — 17.37 руб., 99-й — 27.30 руб.


Набдюдаются следующие аномалии :
- отрицательное числа в  `Quantity`  
- крупное число позиций в `Quantity`
- в колонке `price`  завешенная цена **660** относительно других цен

Неоходимо выявить причину возникновения

-----
-----

#// **Вот тут можно начать анализировть отрицательное количество и дальше думать что с ним делать**

#**//Общий вывод по первому этапу**

я могу написать что я в дальнейшем буду с ними делать, с тем что меня смущает и после этого переходить

----
----
----

#**Этап 2. Предварительный анализ и создание новых признаков**

- Расчет общих метрик (общая выручка, количество заказов, количество клиентов, средний чек).
- Создание признаков: месяц, день недели, год из даты.
- Создание признака: общая сумма заказа (`Price` и `Quantity`).
- Агрегация данных на уровне заказа (`TransactionNo`) для анализа корзины.

In [37]:
# Общая выручка
total_revenue = df['TotalPrice'].sum()
print(f"Общая выручка: {total_revenue:.2f} руб.")

Общая выручка: 62781386.54 руб.


In [38]:
# Общее количество уникальных заказов
total_orders = df['TransactionNo'].nunique()
print(f"Общее количество заказов: {total_orders}")

Общее количество заказов: 19790


In [39]:
# Средний чек (общая выручка / общее количество заказов)
average_order_value = total_revenue / total_orders
print(f"Средний чек: {average_order_value:.2f} руб.")

Средний чек: 3172.38 руб.


### **Создание новых признаков из даты**

Для дальнейшего анализа сезонности и временных трендов, из колонки `Date` будут извлечены год, месяц и день недели.

In [40]:
# Извлечение года, месяца и дня недели
df['Year'] = df['Date'].dt.year
df['Month'] = df['Date'].dt.month
df['DayOfWeek'] = df['Date'].dt.dayofweek # Понедельник=0, Воскресенье=6

display(df.head())

,TransactionNo,Date,ProductNo,ProductName,Price,CustomerNo,Country,Quantity,TotalPrice,Year,Month,DayOfWeek
0,536365,2018-12-01,21730,Glass Star Frosted T-Light Holder,14.61,17850,United Kingdom,6,87.66,2018,12,5
1,536365,2018-12-01,22752,Set 7 Babushka Nesting Boxes,18.09,17850,United Kingdom,2,36.18,2018,12,5
2,536365,2018-12-01,71053,White Moroccan Metal Lantern,13.72,17850,United Kingdom,6,82.32,2018,12,5
3,536365,2018-12-01,84029E,Red Woolly Hottie White Heart,13.72,17850,United Kingdom,6,82.32,2018,12,5
4,536365,2018-12-01,84029G,Knitted Union Flag Hot Water Bottle,13.72,17850,United Kingdom,6,82.32,2018,12,5


In [42]:
# Признак 'TotalPrice' (общая сумма заказа) уже был создан ранее в ячейке e4148efb
# df['TotalPrice'] = df['Quantity'] * df['Price']

# Если вы хотите убедиться, что он существует, вот первые несколько строк DataFrame:
display(df[['Quantity', 'Price', 'TotalPrice']].head())

,Quantity,Price,TotalPrice
0,6,14.61,87.66
1,2,18.09,36.18
2,6,13.72,82.32
3,6,13.72,82.32
4,6,13.72,82.32


### **Агрегация данных на уровне заказа для анализа корзины**

Для анализа содержимого корзины и среднего чека, агрегируем данные по `TransactionNo`, чтобы получить общую сумму и количество уникальных товаров в каждом заказе.

In [41]:
# Агрегируем данные на уровне заказа
orders_df = df.groupby('TransactionNo').agg(
    TotalOrderQuantity=('Quantity', 'sum'),
    TotalOrderPrice=('TotalPrice', 'sum'),
    NumUniqueProducts=('ProductNo', 'nunique'),
    OrderDate=('Date', 'first'),
    CustomerNo=('CustomerNo', 'first'),
    Country=('Country', 'first')
).reset_index()

# Добавляем также признак 'Number of Items' для анализа количества позиций в заказе
orders_df['NumberOfItems'] = df.groupby('TransactionNo')['ProductNo'].count().values

# Извлекаем год, месяц, день недели для агрегированного датафрейма
orders_df['Year'] = orders_df['OrderDate'].dt.year
orders_df['Month'] = orders_df['OrderDate'].dt.month
orders_df['DayOfWeek'] = orders_df['OrderDate'].dt.dayofweek


display(orders_df.head())

,TransactionNo,TotalOrderQuantity,TotalOrderPrice,NumUniqueProducts,OrderDate,CustomerNo,Country,NumberOfItems,Year,Month,DayOfWeek
0,536365,40,552.52,7,2018-12-01,17850,United Kingdom,7,2018,12,5
1,536366,12,145.80,2,2018-12-01,17850,United Kingdom,2,2018,12,5
2,536367,83,1136.34,12,2018-12-01,13047,United Kingdom,12,2018,12,5
3,536368,15,225.54,4,2018-12-01,13047,United Kingdom,4,2018,12,5
4,536369,3,49.05,1,2018-12-01,13047,United Kingdom,1,2018,12,5


# **Этап 3. Исследовательский анализ данных (EDA)**

*   Анализ динамики продаж: выручка и количество заказов по месяцам, сезонность.
*   Анализ товаров: топ-10 по выручке, топ-10 по количеству продаж, распределение цен.
*   Анализ корзины: среднее количество товаров в заказе, средний чек, распределение заказов по сумме.
*   Анализ клиентов: топ-10 по сумме трат, топ-10 по частоте покупок, новые, повторные клиенты.
*   Географический анализ: выручка и количество заказов по странам, средний чек по странам.
*   Анализ ценовых сегментов: зависимость количества продаж от цены.

In [33]:
#топ-10 дорогих товаров

top_price = df.groupby('ProductName')['Price'].max().sort_values(ascending=False).head(10)
print(top_price)

ProductName
Vintage Red Kitchen Cabinet          660.62
Vintage Blue Kitchen Cabinet         660.62
Love Seat Antique White Metal        210.12
Rustic Seventeen Drawer Sideboard    179.37
Regency Mirror With Shutters         179.37
Chest Natural Wood 20 Drawers        138.38
Hall Cabinet With 3 Drawers          112.75
Sweetheart Cream Steel Table Rect    112.71
Cinderella Chandelier                112.71
Ivory Embroidered Quilt               94.81
Name: Price, dtype: float64


# **Этап 4. Ключевые выводы и рекомендации**
*   Формирование основных инсайтов по каждому блоку анализа.
*   Разработка практических рекомендаций для увеличения выручки, удержания клиентов и оптимизации ассортимента.

### **Расчет общей выручки по странам**

Для определения наиболее прибыльных стран, необходимо рассчитать общую выручку для каждой страны. Это поможет выявить ключевые рынки, на которых сосредоточены основные продажи.

In [35]:
# Расчет TotalPrice для каждой строки
df['TotalPrice'] = df['Quantity'] * df['Price']

# Группировка по стране и суммирование TotalPrice
revenue_by_country = df.groupby('Country')['TotalPrice'].sum().sort_values(ascending=False).reset_index()

display(revenue_by_country.head(10))

,Country,TotalPrice
0,United Kingdom,52346877.60
1,Netherlands,2151553.59
2,EIRE,1711819.39
3,Germany,1369839.62
4,France,1329903.39
5,Australia,995414.01
6,Sweden,401879.89
7,Switzerland,361691.96
8,Japan,293155.44
9,Spain,280843.80
